### PR TITLE
internal/docs: document compound IDs decision

### PR DESCRIPTION
**Pull request by zombiezen**
Updates #1860

### CODE REVIEW

File: internal/docs/design.md [hunk 1]
```diff
@@ -204,6 +204,27 @@ needed). However, this would result in inconsistent names over time (e.g., some
...
+
+There are two broad compound ID usage patterns we have observed:
+
+1. Applications will keep resources in the same location or project, so the
+   application will build the ID from the subcomponents rather than passing the
+   entire resource ID around.
+2. Applications will pass a verbatim string from configuration down to the
+   API, since this is what was easily copy-pasteable from the cloud console UI.
+
+Go CDK constructors that take in compound IDs should take in a `string` with the
```
Comment (vangent) [inline internal/docs/design.md]: Nit: add "and URLOpeners".

Comment (zombiezen) [reply internal/docs/design.md]: But URLOpeners take in a `*url.URL`, not a `string`. Can you elaborate on the change you want to see here?

File: internal/docs/design.md [hunk 2]
```diff
@@ -204,6 +204,27 @@ needed). However, this would result in inconsistent names over time (e.g., some
...
+### Compound IDs
+
+Many cloud providers have resource IDs that are made up of subcomponents in
+some well-defined syntax. For example, [GCP KMS key IDs][] take the form
+`projects/[PROJECT_ID]/locations/[LOCATION]/keyRings/[KEY_RING]/cryptoKeys/[KEY]`.
+We call these _compound IDs_.
+
+There are two broad compound ID usage patterns we have observed:
+
+1. Applications will keep resources in the same location or project, so the
```
Comment (eliben) [inline internal/docs/design.md]: Nit: not sure what the first part of the sentence (up to "project") means and how it helps explain the pattern. Consider dropping it

File: internal/docs/design.md (hunk 1 above)
Comment (vangent) [reply internal/docs/design.md]: I mean that the URL used to open the resource should have the full compound ID in them. So, in your example, the URL should have `projects/[PROJECT_ID]/locations/[LOCATION]/keyRings/[KEY_RING]/cryptoKeys/[KEY]` in it somewhere.

File: internal/docs/design.md (hunk 2 above)
Comment (zombiezen) [reply internal/docs/design.md]: Done.

File: internal/docs/design.md (hunk 1 above)
Comment (zombiezen) [reply internal/docs/design.md]: Gotcha. Added a slightly different wording to be more precise, PTAL.

Comment (vangent) [reply internal/docs/design.md]: SGTM.

**Pull request merged**